vm/systemSmartContracts/esdt.go

```diff
@@ -0,0 +1,279 @@
+type ArgsNewESDTSmartContract struct {
+	Eei           vm.SystemEI
+	GasCost       vm.GasCost
+	ESDTSCConfig  config.ESDTSystemSCConfig
+	ESDTSCAddress []byte
+	Marshalizer   marshal.Marshalizer
+	Hasher        hashing.Hasher
+}
+
+// NewESDTSmartContract create the esdt smart contract, which controls the issuing of tokens
+func NewESDTSmartContract(args ArgsNewESDTSmartContract) (*esdt, error) {
+	if check.IfNil(args.Eei) {
+		return nil, vm.ErrNilSystemEnvironmentInterface
+	}
+	if check.IfNil(args.Marshalizer) {
+		return nil, vm.ErrNilMarshalizer
+	}
+	if check.IfNil(args.Hasher) {
+		return nil, vm.ErrNilHasher
+	}
+
+	baseIssuingCost, ok := big.NewInt(0).SetString(args.ESDTSCConfig.BaseIssuingCost, conversionBase)
+	if !ok || baseIssuingCost.Cmp(big.NewInt(0)) < 0 {
+		return nil, vm.ErrInvalidBaseIssuingCost
+	}
+
+	return &esdt{
+		eei:             args.Eei,
+		gasCost:         args.GasCost,
+		baseIssuingCost: baseIssuingCost,
+		ownerAddress:    []byte(args.ESDTSCConfig.OwnerAddress),
+		eSDTSCAddress:   args.ESDTSCAddress,
+		hasher:          args.Hasher,
+		marshalizer:     args.Marshalizer,
+	}, nil
+}
+
+// Execute  calls one of the functions from the esdt smart contract and runs the code according to the input
+func (e *esdt) Execute(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if CheckIfNil(args) != nil {
+		return vmcommon.UserError
+	}
+
+	switch args.Function {
+	case "_init":
+		return e.init(args)
+	case "issue":
+		return e.issue(args)
+	case "issueProtected":
+		return e.issueProtected(args)
+	case "burn":
+		return e.burn(args)
+	case "mint":
+		return e.mint(args)
+	case "freeze":
+		return e.freeze(args)
+	case "wipe":
+		return e.wipe(args)
+	case "pause":
+		return e.pause(args)
+	case "unPause":
+		return e.unpause(args)
+	case "claim":
+		return e.claim(args)
+	case "configChange":
+		return e.configChange(args)
+	case "esdtControlChanges":
+		return e.esdtControlChanges(args)
+	}
+
+	return vmcommon.Ok
+}
+
+func (e *esdt) init(_ *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	scConfig := &ESDTConfig{
+		OwnerAddress:       e.ownerAddress,
+		BaseIssuingCost:    e.baseIssuingCost,
+		MinTokenNameLength: minLengthForTokenName,
+		MaxTokenNameLength: maxLengthForTokenName,
+	}
+	marshalledData, err := e.marshalizer.Marshal(scConfig)
+	log.LogIfError(err, "marshall error on esdt init function")
+
+	e.eei.SetStorage([]byte(configKeyPrefix), marshalledData)
+	return vmcommon.Ok
+}
+
+func (e *esdt) issueProtected(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if len(args.Arguments) < 3 {
+		return vmcommon.FunctionWrongSignature
+	}
+	if len(args.Arguments[0]) < len(args.CallerAddr) {
+		return vmcommon.FunctionWrongSignature
+	}
+	if args.CallValue.Cmp(e.baseIssuingCost) != 0 {
+		return vmcommon.OutOfFunds
+	}
+	err := e.eei.UseGas(e.gasCost.MetaChainSystemSCsCost.ESDTIssue)
+	if err != nil {
+		return vmcommon.OutOfGas
+	}
+
+	err = e.issueToken(args.Arguments[0], args.Arguments[1:])
```

Comment:
I think I can impersonate the centralized issuing address here. Why not checking that args.CallerAddr is the permitted address got from the config? In this way issueProtected is not needed anymore.

Reply:
protected is not protected to be called only by the owner address - good catch